sound/soc/sof/topology.c

```diff
@@ -2462,15 +2463,161 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
 	.bytes_ext_ops_count = ARRAY_SIZE(sof_dspless_bytes_ext_ops),
 };
 
+#define MAX_TPLG_NUM 5
+
+enum tplg_device_id {
+	TPLG_DEVICE_SDW_JACK,
+	TPLG_DEVICE_SDW_AMP,
+	TPLG_DEVICE_SDW_MIC,
+	TPLG_DEVICE_HOST_DMIC,
+	TPLG_DEVICE_HDMI,
+};
+
+struct topology_file {
+	char *device;
+	char *file;
+	int be_id;
+};
+
+static bool is_platform_support_separated_tplg(const char *platform)
+{
+
+	if (!strcmp(platform, "hda") || !strcmp(platform, "mtl") ||
+	    !strcmp(platform, "lnl") || !strcmp(platform, "ptl"))
+		return true;
+
+	return false;
+}
+
+/* The code is from https://stackoverflow.com/questions/47116974/remove-a-substring-from-a-string-in-c */
+static char *strremove(char *str, const char *sub) {
+	size_t len = strlen(sub);
+	if (len > 0) {
+		char *p = str;
+		size_t size = 0;
+		while ((p = strstr(p, sub)) != NULL) {
+			size = (size == 0) ? (p - str) + strlen(p + len) + 1 : size - len;
+			memmove(p, p + len, size - (p - str));
+		}
+	}
+	return str;
+}
+
 int snd_sof_load_topology(struct snd_soc_component *scomp, const char *file)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_pdata *sof_pdata = sdev->pdata;
+	struct topology_file tplg_files[MAX_TPLG_NUM];
+	struct snd_soc_dai_link *dai_link;
 	const struct firmware *fw;
+	bool load_default_tplg = false;
+	unsigned long tplg_mask = 0;
+	char platform[4];
+	char *tplg_name;
+	int tplg_num = 0;
+	int tplg_dev;
 	int ret;
+	int i;
 
 	dev_dbg(scomp->dev, "loading topology:%s\n", file);
 
-	ret = request_firmware(&fw, file, scomp->dev);
+	tplg_name = (char *)file;
+	if (sdev->pdata->ipc_type == SOF_IPC_TYPE_3)
+		goto legacy_tplg;
+
+	sscanf(sof_pdata->tplg_filename, "sof-%3s-*.tplg", platform);
+	if (!is_platform_support_separated_tplg(platform))
+		goto legacy_tplg;
+
+	for_each_card_prelinks(scomp->card, i, dai_link) {
+		char *tplg_device;
+
+		dev_dbg(scomp->dev, "dai_link %s id %d\n", dai_link->name, dai_link->id);
+		if (strstr(dai_link->name, "SimpleJack")) {
+			tplg_dev = TPLG_DEVICE_SDW_JACK;
+			tplg_device = "sdca-jack";
+		} else if (strstr(dai_link->name, "SmartAmp")) {
+			tplg_dev = TPLG_DEVICE_SDW_AMP;
+			tplg_device = devm_kasprintf(sdev->dev, GFP_KERNEL,
+						     "sdca-%damp", dai_link->num_cpus);
+		} else if (strstr(dai_link->name, "SmartMic")) {
+			tplg_dev = TPLG_DEVICE_SDW_MIC;
+			tplg_device = "sdca-mic";
```

Comment:
Here and for the other checks: we need to update this match list every time a new component got introduced to products and we have new tplg fragment?
Is it going to scale?

Reply:
We need to update this match list when a new dai links got introduced. Given that we didn't update dai links in our machine drivers. I think we will only update the match list occasionally.